src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.DecisionDagRewriter.cs

```diff
@@ -137,13 +137,39 @@ public override BoundNode VisitCall(BoundCall node)
                     bool mightMutate =
                         // might be a call to a local function that assigns something
                         node.Method.MethodKind == MethodKind.LocalFunction ||
-                        // or perhaps we are passing a variable by ref and mutating it that way
-                        !node.ArgumentRefKindsOpt.IsDefault;
+                        // or perhaps we are passing a variable by ref and mutating it that way, e.g. `int.Parse(..., out x)`
+                        !node.ArgumentRefKindsOpt.IsDefault ||
+                        // or perhaps we are calling a mutating method of a value type
+                        MethodMayMutateReceiver(node.ReceiverOpt, node.Method);
 
                     if (mightMutate)
                         _mightAssignSomething = true;
                     else
                         base.VisitCall(node);
+                    return null;
+                }
+
+                private static bool MethodMayMutateReceiver(BoundExpression receiver, MethodSymbol method)
+                {
+                    return
+                        method != null &&
+                        !method.IsStatic &&
+                        !method.IsEffectivelyReadOnly &&
```

Comment:
>!method.IsEffectivelyReadOnly && [](start = 24, length = 32)

I might be wrong but it doesn't look like any tests were added to confirm/verify significance of this check. What about other conditions in this function?

Reply:
Other conditions are exercised by existing tests.  I will add a test that demonstrates (using unsafe) that the optimization causes incorrect code to be generated when the `readonly` modifier is not obeyed by the method's implementation, thus demonstrating that the optimization has been applied.

---
In reply to: [405675096](https://github.com/dotnet/roslyn/pull/42972#discussion_r405675096) [](ancestors = 405675096)

Reply:
> I will add a test that demonstrates (using unsafe) that the optimization causes incorrect code to be generated when the readonly modifier is not obeyed by the method's implementation, thus demonstrating that the optimization has been applied.

Perhaps it could be observed by observing a difference in code gen when a method is changed from readonly to not-readonly, or vise versa.  


---
In reply to: [405839405](https://github.com/dotnet/roslyn/pull/42972#discussion_r405839405) [](ancestors = 405839405,405675096)

Reply:
Sure, I will do it that way, but such tests are generally hard to review and maintain.

---
In reply to: [405841250](https://github.com/dotnet/roslyn/pull/42972#discussion_r405841250) [](ancestors = 405841250,405839405,405675096)